### DISCUSSION TITLE
Demo console app - issue with stream not progressing (different issue)

### DISCUSSION THREAD

**Discussion by InTheta** [psollberger/DeriSock#53] [Q&A] [answered]
resolved

**Comment by psollberger** [comment]
Unfortunately I don't have enough context to help you out with this right now.
Can you show me the full code? Particularly the implementation of `Deribit.Core.Deribit` which isn't from DeriSock

If you want to provide it as a gist, please make sure to name the file with the `.cs` extension so it get's code highlighting :)

**Comment by psollberger** [comment, marked as answer]
Your code in (I guess) the console app ...
```cs
await Setup();
Subscribe(); // not awaited
ExitEvent.WaitOne();
```
... is wrong. The `// not awaited` part is wrong.

This works:
```cs
await Setup();
await Subscribe();
```

Your usage of async/await somehow produces a dead-lock.

**Comment by InTheta** [comment]


**Reply by psollberger** [reply]
I added a little simplification as a comment in the gist.

As for "What is the best practice for this?": I don't really have an answer to that. It's async/await and therefore highly depends.
You need to understand what async/await does under the hood and how ConfigureAwait(false) changes the behavior.

I can only suggest to use ConfigureAwait(false) whenever you don't need to continue on the same thread the execution was in when it startet the task. In a console application you nearly never need to continue on the same thread.

**Reply by psollberger** [reply]
In my opinion, the job of async/await is asynchroneous execution, not parallel execution.

If you need parallel execution I suggest to use a distinct Thread or a long running Task.

In fact, DeribitClient uses a LongRunning Task (although i thought about replacing it with a distinct thread) to handle incoming messages.
See:
- Task creation: https://github.com/psollberger/DeriSock/blob/main/src/DeriSock/DeribitClient.cs#L116
- Function run in that task: https://github.com/psollberger/DeriSock/blob/main/src/DeriSock/DeribitClient.cs#L172
